### PR TITLE
tickets/DM-27800: Add option to perform exact match on catalogs

### PR DESCRIPTION
**Pull request by fred3m**
This PR exists to implement the exact source matching from DM-27800. It is a placeholder for the work using afw catalogs, but will not be merged due to the upcoming switch from afw tables to pandas, which will require another ticket.

### CODE REVIEW

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 1]
```diff
@@ -1613,6 +1615,11 @@ def getTargetList(parsedCmd, **kwargs):
                 refList1, refList2 in zip(parsedCmd.id.refList, idParser.refList)]
 
 
+def patch2Int(patch):
```
Comment (PaulPrice) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: I believe the use of `2` in the name like this is discouraged; use `To`.

Add docstring.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 2]
```diff
@@ -1613,6 +1615,11 @@ def getTargetList(parsedCmd, **kwargs):
...
+    assert len(patch) == 3
```
Comment (PaulPrice) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: Where does this limitation come from? I work on tracts that have a patch like `15,20`, which would violate this.

I think you should do:

```
def patchToInt(patch, maxPatches):
    """docstring"""
    elements = patch.split(",")
    assert len(elements) == 2
    return maxPatches*int(elements[0]) + int(elements[1])
```

(You can get `maxPatches` from the skyMap, or make it a config parameter if you're desperate. Hard-coding a value will cause problems later.)

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 3]
```diff
@@ -1757,6 +1764,7 @@ def runDataRef(self, patchRefList1, patchRefList2, subdir=""):
         else:
             forced1 = unforced1
             forced2 = unforced2
+
```
Comment (PaulPrice) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: You shouldn't add lines that aren't related to the stated function of the commit.

File: python/lsst/pipe/analysis/coaddAnalysis.py [hunk 4]
```diff
@@ -67,6 +67,8 @@ class CoaddAnalysisConfig(Config):
     coaddName = Field(dtype=str, default="deep", doc="Name for coadd")
     matchRadiusRaDec = Field(dtype=float, default=0.5, doc="RaDec Matching radius (arcseconds)")
     matchOverlapRadius = Field(dtype=float, default=0.5, doc="Matching radius for overlaps (arcseconds)")
+    matchExact = Field(dtype=bool, default=False,
+                       doc="Match catalogs derived from the same mergeDet catalog exactly")
```
Comment (PaulPrice) [inline python/lsst/pipe/analysis/coaddAnalysis.py]: Add a `?` to the end of the `doc` string to reflect the `bool` nature.

**Review by PaulPrice — CHANGES_REQUESTED** [review]
I think there are a few things to change before this is merged.

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 2 above)
Comment (fred3m) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: Definitely. Good point.

File: python/lsst/pipe/analysis/coaddAnalysis.py (hunk 3 above)
Comment (fred3m) [reply python/lsst/pipe/analysis/coaddAnalysis.py]: I think that I added a temporary debugging statement here and left the code in the way that I would have written it, with an extra space. I'll remove.

**Comment by fred3m** [comment]
This PR is closed since it was never intended to be merged (see the first comment above), as pipe_tasks will no longer use afw catalogs once DM-22266 is merged in the next week or so. This branch exists to store the work necessary to test meas_extensions_scarlet in the time before DM-22266 is merged. Once that merge occurs, a new ticket will re-implement exact matching.

In regards to choosing to implement the `patchToInt` function as a config option instead of calculating it directly, the gen3 Butler stores all patch ID's as integers, so that function is only needed until pipe_analysis uses the gen3 middleware.

**Pull request closed without merging**